presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java

```diff
@@ -1068,12 +1072,30 @@ private void validateStorageCapabilities(TempStorage tempStorage)
         private void queryCompletedEvent(Optional<ExecutionFailureInfo> failureInfo, OptionalLong updateCount)
         {
             List<SerializedTaskInfo> serializedTaskInfos = taskInfoCollector.value();
-            List<TaskInfo> taskInfos = serializedTaskInfos.stream()
-                    .map(SerializedTaskInfo::getBytes)
-                    .map(PrestoSparkUtils::decompress)
-                    .map(taskInfoJsonCodec::fromJson)
-                    .collect(toImmutableList());
-            StageInfo stageInfo = createStageInfo(session.getQueryId(), fragmentedPlan, taskInfos);
+            ImmutableList.Builder<TaskInfo> taskInfos = ImmutableList.builder();
+            long totalSerializedTaskInfoSizeInBytes = 0;
+            boolean missingPipelineStats = false;
+            Set<Integer> fragmentsWithMissingPipelineStats = new HashSet<>();
+            for (SerializedTaskInfo serializedTaskInfo : serializedTaskInfos) {
+                byte[] bytes = serializedTaskInfo.clear();
+                totalSerializedTaskInfoSizeInBytes += bytes.length;
+                TaskInfo taskInfo = deserializeZstdCompressed(taskInfoCodec, bytes);
+                if (serializedTaskInfo.getCompactionLevel().isIncomplete()) {
+                    missingPipelineStats = true;
+                    fragmentsWithMissingPipelineStats.add(taskInfo.getTaskId().getStageExecutionId().getStageId().getId());
+                }
+                taskInfos.add(taskInfo);
+            }
+            taskInfoCollector.reset();
+
+            log.info("Total serialized task info size: %s", DataSize.succinctBytes(totalSerializedTaskInfoSizeInBytes));
+            if (missingPipelineStats) {
+                String message = format("Missing pipeline statistics for fragments %s, operator level metrics may be incomplete", fragmentsWithMissingPipelineStats);
+                log.warn(message);
+                warningCollector.add(new PrestoWarning(INCOMPLETE_OPERATOR_STATISTICS, message));
```

Comment:
> This is defensive? We don't generally expect to hit this?

Yeah, this should be defensive. We don't expect significant number of queries to hit this limit. 

> I wonder how many external integrations are going to be thrown off because they mine incomplete statistics and are unaware of the warning.

We don't drop high level statistics. We only drop operator level statistics that are used to annotate plan with "per plan node" execution statistics and to store operator level statistics. Higher level statistics shouldn't be affected by this change.

Reply:
My concern is they get used for something like HBO (@ahmadghazal64) and having an incomplete data set might throw them off?

The case I want to make is if we want to have a lossy statistics column we may want to go as far as creating a lossy statistics column and only populate the current statistics column if it's complete.

Reply:
Discussed offline, decided to drop this optimization for now